firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java

```diff
@@ -205,13 +210,68 @@ private void recordFirstDrawDone() {
     if (firstDrawDone != null) {
       return;
     }
+    Timer start = getStartTimer();
     this.firstDrawDone = clock.getTime();
-    executorService.execute(
-        () -> this.logColdStart(getStartTimer(), this.firstDrawDone, this.startSession));
+    int subtraceCount = this.experimentTtid.getSubtracesCount();
+    this.experimentTtid
+        .setClientStartTimeUs(start.getMicros())
+        .setDurationUs(start.getDurationMicros(this.firstDrawDone));
 
-    if (isRegisteredForLifecycleCallbacks) {
-      // After AppStart trace is queued to be logged, we can unregister this callback.
-      unregisterActivityLifecycleCallbacks();
+    TraceMetric.Builder subtrace =
+        TraceMetric.newBuilder()
+            .setName("_experiment_classLoadTime")
+            .setClientStartTimeUs(FirebasePerfProvider.getAppStartTime().getMicros())
+            .setDurationUs(
+                FirebasePerfProvider.getAppStartTime().getDurationMicros(this.firstDrawDone));
+    this.experimentTtid.addSubtraces(subtrace.build());
+
+    subtrace = TraceMetric.newBuilder();
+    subtrace
+        .setName("_experiment_uptimeMillis")
+        .setClientStartTimeUs(start.getMicros())
+        .setDurationUs(start.getDurationUptimeMicros(this.firstDrawDone));
+    this.experimentTtid.addSubtraces(subtrace.build());
+
+    this.experimentTtid.addPerfSessions(this.startSession.build());
+
+    if (subtraceCount > 0) {
+      executorService.execute(() -> this.logExperimentTtid(this.experimentTtid));
+
+      if (isRegisteredForLifecycleCallbacks) {
+        // After AppStart trace is queued to be logged, we can unregister this callback.
+        unregisterActivityLifecycleCallbacks();
+      }
+    }
+  }
+
+  private void recordFirstDrawDonePreDraw() {
+    if (preDraw != null) {
+      return;
+    }
+    Timer start = getStartTimer();
+    this.preDraw = clock.getTime();
+    int subtraceCount = this.experimentTtid.getSubtracesCount();
+    TraceMetric.Builder subtrace =
+        TraceMetric.newBuilder()
+            .setName("_experiment_preDraw")
+            .setClientStartTimeUs(start.getMicros())
+            .setDurationUs(start.getDurationMicros(this.preDraw));
+    this.experimentTtid.addSubtraces(subtrace.build());
+
+    subtrace = TraceMetric.newBuilder();
+    subtrace
+        .setName("_experiment_preDraw_uptimeMillis")
+        .setClientStartTimeUs(start.getMicros())
+        .setDurationUs(start.getDurationUptimeMicros(this.preDraw));
+    this.experimentTtid.addSubtraces(subtrace.build());
+
+    if (subtraceCount > 0) {
+      executorService.execute(() -> this.logExperimentTtid(this.experimentTtid));
```

Comment:
What does it mean to log this.experimentTtid twice?

Once in `recordFirstDrawDone` with 2 subtraces and then `recordFirstDrawDonePreDraw` with 2 more (total of 4)?

Reply:
This if block of code is because I am not sure if the order with which the "`postAtFrontOfQueue` in `onDraw`" and the "`post` in `onPreDraw`" is deterministic. `subtraceCount` is taken at the beginning of the method, so if the other method already was called, it would not be zero, so we log the trace. However, if the other method was not called, then the current method is the first one to be invoked, it shouldn't log the trace because it still needs to wait for the other method.

Reply:
I see. It would be more clear to use either a class variable like hasLoggedExperiment and a comment to make it more obvious what is being done here.